File: src/engine/statistics/correlation.ts
import jStat from 'jstat';
import type { AnalysisResult } from '../../types';

/**
 * Pearson correlation matrix for multiple variables (one row per case, one column per variable).
 * @spec PRISM-TBL-009
 */
export function runCorrelation(
  variableLabels: string[],
  rows: (number | null)[][]
): AnalysisResult {
  const nVars = variableLabels.length;
  const r: number[][] = Array.from({ length: nVars }, () => Array(nVars).fill(0));
  const n: number[][] = Array.from({ length: nVars }, () => Array(nVars).fill(0));

  for (let i = 0; i < nVars; i++) {
    r[i][i] = 1;
    n[i][i] = rows.filter((row) => {
      const v = row[i];
      return v != null && Number.isFinite(v);
    }).length;
    for (let j = i + 1; j < nVars; j++) {
      const pairs: { x: number; y: number }[] = [];
      for (let row = 0; row < rows.length; row++) {
        const xi = rows[row][i];
        const yj = rows[row][j];
        if (xi != null && Number.isFinite(xi) && yj != null && Number.isFinite(yj)) {
          pairs.push({ x: xi, y: yj });
        }
      }
      const nn = pairs.length;
      n[i][j] = nn;
      n[j][i] = nn;
      if (nn < 2) {
        r[i][j] = NaN;
        r[j][i] = NaN;
      } else {
        const cor = (jStat as unknown as { corrcoeff(a: number[], b: number[]): number }).corrcoeff(
          pairs.map((p) => p.x),
          pairs.map((p) => p.y)
        );
        r[i][j] = cor;
        r[j][i] = cor;
      }
    }
  }

  return {
    type: 'correlation',
    labels: variableLabels,
    r,
    n,
  };
}

File: backend/app/engine/correlation.py
"""Correlation analysis."""

from __future__ import annotations

import numpy as np
from scipy import stats

from backend.app.models import CorrelationResult


def run_correlation(
    labels: list[str],
    rows: list[list[float | None]],
) -> CorrelationResult:
    """Compute correlation matrix.

    Args:
        labels: Variable labels
        rows: Data rows (one row per observation, one column per variable)

    Returns:
        CorrelationResult with correlation matrix
    """
    n_vars = len(labels)

    cols = []
    for col_idx in range(n_vars):
        values = [r[col_idx] for r in rows if r[col_idx] is not None and not np.isnan(r[col_idx])]
        cols.append(values)

    min_len = min(len(c) for c in cols) if cols else 0
    if min_len < 3:
        raise ValueError("Correlation requires at least 3 complete observations")

    data = np.array(
        [
            [r[col_idx] if r[col_idx] is not None else np.nan for col_idx in range(n_vars)]
            for r in rows
        ]
    )

    mask = ~np.isnan(data).any(axis=1)
    clean_data = data[mask]

    if len(clean_data) < 3:
        raise ValueError("Correlation requires at least 3 complete observations")

    corr_matrix = np.corrcoef(clean_data.T)

    return CorrelationResult(
        type="correlation",
        matrix=corr_matrix.tolist(),
        labels=labels,
        method="pearson",
    )
